python/ray/dataframe/dataframe.py

```diff
@@ -2849,34 +2848,57 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
-                return df.quantile(q=q, axis=axis, numeric_only=numeric_only,
-                                   interpolation=interpolation)
-            except ValueError:
-                return pd.Series()
+        # deal with error checking for bad dtyping when numeric_only is False
+        if not numeric_only:
+            # check if there are any object columns
+            for t in self.dtypes:
+                if t == np.dtype('O'):
+                    e_msg = "Cannot call quantile on columns of type " + str(t)
+                    raise TypeError(e_msg)
+
+        # check that all qs are between 0 and 1
+        pd.DataFrame()._check_percentile(q)
 
         if isinstance(q, (pd.Series, np.ndarray, pd.Index, list)):
-            # In the case of a list, we build it one at a time.
-            # TODO Revisit for performance
-            quantiles = []
-            for q_i in q:
-                def remote_func(df):
-                    return quantile_helper(df, q=q_i, axis=axis,
-                                           numeric_only=numeric_only,
-                                           interpolation=interpolation)
-
-                result = self._arithmetic_helper(remote_func, axis)
-                result.name = q_i
-                quantiles.append(result)
-
-            return pd.concat(quantiles, axis=1).T
+
+            if axis == 0 or axis == 'index':
+                result = _map_partitions(
+                                        lambda df: df.quantile(
+                                                q, axis, numeric_only,
+                                                interpolation),
+                                        self._col_partitions)
+                q_index = pd.Float64Index(q)
+                return DataFrame(col_partitions=result,
+                                 index=q_index,
+                                 columns=self.columns)
+
+            if axis == 1 or axis == 'columns':
+                result = _map_partitions(
+                                        lambda df: df.quantile(
+                                                q, axis, numeric_only,
+                                                interpolation),
+                                        self._row_partitions)
+                q_index = pd.Float64Index(q)
+                return DataFrame(col_partitions=result,
+                                 index=q_index,
+                                 columns=self.index)
+
+        # single q means we can use arithmetic helper
         else:
+            def quantile_helper(df, q, axis, numeric_only, interpolation):
+                try:
+                    return df.quantile(q=q, axis=axis,
+                                       numeric_only=numeric_only,
+                                       interpolation=interpolation)
+                # This exception is thrown when there are only non-numeric columns
+                # in this partition
+                except ValueError:
```

Comment:
As discussed above, `TypeErrors` are thrown here. Perhaps you can add some logic at the beginning of the function to ensure all the `dtypes` are comparable. 

The issue seems to arise not from non-numeric data, but from non-comparable types.